File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) { // O(1) memory solution
    int out = 0;
    if (!height.size())
        return 0;
    for(int L = 0, R = height.size() - 1, maxL = height[L], maxR = height[R], sum = 0;L < R;)
    {
        if(maxL < maxR)
        {
            L++;
            maxL = max(maxL, height[L]);
            sum = min(maxL,maxR) - height[L];
        }
        else
        {
            R--;
            maxR = max(maxR, height[R]);
            sum = min(maxL,maxR) - height[R];
        }
        if( sum > 0)
            out += sum;
    }
    return out;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        if not n: return 0
        leftMax = [0] * n
        rightMax = [0] * n

        leftMax[0] = height[0]

        for i in range(1 , n):
            leftMax[i] = max(leftMax[i - 1], height[i])
        
        rightMax[n - 1] = height[n - 1]

        for i in range(n - 2, 0, -1):
            rightMax[i] = max(rightMax[i + 1], height[i])

        out = 0
        for i in range(1, n):
            out += min(leftMax[i], rightMax[i]) - height[i]
        return out
       